File: CoreData/CoreData/ViewController.m
//
//  ViewController.m
//  CoreData
//
//  Created by zhangpei on 16/6/15.
//  Copyright © 2016年 ivan. All rights reserved.
//

#import "ViewController.h"

#import <CoreData/CoreData.h>
#import "Dog.h"
#import "AppDelegate.h"


@interface ViewController ()
{
    AppDelegate * app;
}

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    app = [[UIApplication sharedApplication] delegate];
    
    [self createButton];
    
}


-(void)createButton
{
    
    NSArray * arr = [NSArray arrayWithObjects: @"增", @"删", @"改", @"查", nil ];
    
    NSInteger btnY = 50;
    for ( NSInteger i = 0 ; i < arr.count; i++ ) {
        UIButton * btn = [UIButton buttonWithType: UIButtonTypeCustom ];
        btn.frame = CGRectMake( 10, btnY, 100, 100 );
        btn.backgroundColor = [UIColor redColor];
        btn.tag = 100 + i;
        [btn addTarget: self action:@selector(clickBtn:) forControlEvents:UIControlEventTouchUpInside ];
        [btn setTitle: [NSString stringWithFormat:@"%@", [arr objectAtIndex: i ]] forState:UIControlStateNormal ];
        [self.view addSubview: btn ];
        btnY += btn.frame.size.height + 10;
    }
    
    
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(void)clickBtn:(UIButton *)btn
{
    switch ( btn.tag ) {
        case 100:
        {
            [self clickBtnAdd];
            
            break;
        }
        case 101:
        {
            [self clickBtnDelete];
            
            break;
        }
        case 102:
        {
            [self clickBtnUpdate];
            
            break;
        }
        case 103:
        {
            [self clickBtnSelect];
            
            break;
        }
        default:
            break;
    }
    
}


-(void)clickBtnAdd
{
    int getAge = arc4random() % 1000;
    
    Dog * dog = [NSEntityDescription insertNewObjectForEntityForName:@"Dog" inManagedObjectContext:app.managedObjectContext ];
    dog.name = [NSString stringWithFormat:@"狗狗%d", getAge ];
    dog.age = [NSString stringWithFormat:@"狗%d", getAge ];
    dog.sex = [NSString stringWithFormat:@"%@狗狗", (getAge % 2) ? @"男" : @"女" ];
    
    [app.managedObjectContext save: nil ];
    
}

-(void)clickBtnDelete
{
    NSEntityDescription * getEntityDescription = [NSEntityDescription entityForName:@"Dog" inManagedObjectContext: app.managedObjectContext ];
    NSFetchRequest * request = [[NSFetchRequest alloc] init];
    [request setEntity: getEntityDescription ];
    
    
    NSPredicate * predicate = [NSPredicate predicateWithFormat:@"name=%@", @"花花" ];
    [request setPredicate: predicate ];
    
    NSArray * getArr = (NSArray *)[app.managedObjectContext executeFetchRequest: request error: nil ];
    
    if ( getArr.count ) {
        for ( Dog * dog in getArr ) {
            NSLog( @"dog--name==%@", dog.name );
            NSLog( @"dog--sex==%@", dog.sex );
            NSLog( @"dog--age==%@", dog.age );
            
            [app.managedObjectContext deleteObject: dog ];
        }
        [app.managedObjectContext save: nil ];
        NSLog( @"删除完成" );
        
    }else{
        NSLog( @" 没有查到 " );
        
    }
}

-(void)clickBtnUpdate
{
    NSEntityDescription * getEntityDescription = [NSEntityDescription entityForName:@"Dog" inManagedObjectContext: app.managedObjectContext ];
    NSFetchRequest * request = [[NSFetchRequest alloc] init];
    [request setEntity: getEntityDescription ];
    
    
    NSPredicate * predicate = [NSPredicate predicateWithFormat:@"name!=%@", @"小花狗" ];
    [request setPredicate: predicate ];
    
    NSArray * getArr = (NSArray *)[app.managedObjectContext executeFetchRequest: request error: nil ];
    
    if ( getArr.count ) {
        for ( Dog * dog in getArr ) {
            NSLog( @"dog--name==%@", dog.name );
            NSLog( @"dog--sex==%@", dog.sex );
            NSLog( @"dog--age==%@", dog.age );
            dog.name = @"小花狗";
            dog.sex = @"0";
            dog.age = @"0";
        }
        [app.managedObjectContext save: nil ];
        NSLog( @"修改完成" );
        
    }else{
        NSLog( @" 没有查到 " );
        
    }
    
}

-(void)clickBtnSelect
{
    NSEntityDescription * getEntityDescription = [NSEntityDescription entityForName:@"Dog" inManagedObjectContext: app.managedObjectContext ];
    NSFetchRequest * request = [[NSFetchRequest alloc] init];
    [request setEntity: getEntityDescription ];
    
    NSArray * getArr = [app.managedObjectContext executeFetchRequest: request error: nil ];
    
    if ( getArr.count != 0 ) {
        
        for ( Dog * dog in getArr ) {
            NSLog( @"\n\ndog--name==%@", dog.name );
            NSLog( @"dog--sex==%@", dog.sex );
            NSLog( @"dog--age==%@", dog.age );
        }
    }
    
}







@end

File: CoreDataSwift/CoreDataSwift/ViewController.swift
//
//  ViewController.swift
//  CoreDataSwift
//
//  Created by zhangpei on 16/6/16.
//  Copyright © 2016年 ivan. All rights reserved.
//


import UIKit
import CoreData

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        createBtn()
        
    }
    
    func createBtn(){
        
        var getArr = [ "增", "删", "改", "查" ];
        
        var btnY:CGFloat = 50;
        
        for var index = 0; index < getArr.count; ++index {
            
            var button:UIButton = UIButton(type:.Custom )
            button.frame = CGRectMake(10, btnY, 100, 30)
            button.tag = 100 + index
            button.backgroundColor = UIColor.redColor()
            button.setTitle( getArr[index], forState:UIControlState.Normal )
            self.view.addSubview(button);
            button.addTarget(self,action:#selector(clickBtn(_:)), forControlEvents:.TouchUpInside)
            
            btnY += button.frame.size.height + 10
            
        }
    }
    
    func clickBtn(btn:UIButton){
        switch ( btn.tag ) {
        case 100:
                clickBtnAdd ();
                break;
            
        case 101:
                clickBtnDelete ();
                break;
            
        case 102:
                clickBtnUpdate ();
                break;
            
        case 103:
                clickBtnSelect ();
                break;
        default:
            break;
        }
    }
    
    
    func clickBtnAdd(){
        let getAge: Int = Int( arc4random() % 1000 );
        let app = UIApplication.sharedApplication().delegate as! AppDelegate;
        let context = app.managedObjectContext
        
        let dog = NSEntityDescription.insertNewObjectForEntityForName( "Dog", inManagedObjectContext:  context ) as! Dog;
        dog.name = "狗";
        dog.age = String(getAge);
        dog.sex = "男";
        
        //保存
        do {
            try context.save()
            print("保存成功！")
        } catch {
            fatalError("不能保存：\(error)")
        }
    }
    
    
    func clickBtnDelete() {
        //获取管理的数据上下文 对象
        let app = UIApplication.sharedApplication().delegate as! AppDelegate
        let context = app.managedObjectContext
        
        //声明数据的请求
        let fetchRequest:NSFetchRequest = NSFetchRequest()
        fetchRequest.fetchLimit = 10 //限定查询结果的数量
        fetchRequest.fetchOffset = 0 //查询的偏移量
        
        //声明一个实体结构
        let entity:NSEntityDescription? = NSEntityDescription.entityForName("Dog", inManagedObjectContext: context)
        //设置数据请求的实体结构
        fetchRequest.entity = entity
        
        //设置查询条件
        let predicate = NSPredicate( format: "name= '狗'" )
        fetchRequest.predicate = predicate
        
        //查询操作
        do {
            let fetchedObjects:[AnyObject]? = try context.executeFetchRequest(fetchRequest)
            
            //遍历查询的结果
            for info:Dog in fetchedObjects as! [Dog]{
                //删除对象
                print("\nname=\(info.name)")
                print("sex=\(info.sex)")
                print("age=\(info.age)")
                
                context.deleteObject(info)
            }
            
            //重新保存-更新到数据库
            try context.save()
            
        }
        catch {
            fatalError("不能保存：\(error)")
        }
    }
    
    func clickBtnUpdate() {
        //获取管理的数据上下文 对象
        let app = UIApplication.sharedApplication().delegate as! AppDelegate
        let context = app.managedObjectContext
        
        //声明数据的请求
        let fetchRequest:NSFetchRequest = NSFetchRequest()
        fetchRequest.fetchLimit = 10 //限定查询结果的数量
        fetchRequest.fetchOffset = 0 //查询的偏移量
        
        //声明一个实体结构
        let entity:NSEntityDescription? = NSEntityDescription.entityForName("Dog", inManagedObjectContext: context)
        //设置数据请求的实体结构
        fetchRequest.entity = entity
        
        //设置查询条件
        let predicate = NSPredicate( format: "sex = '男'" )
        fetchRequest.predicate = predicate
        
        //查询操作
        do {
            let fetchedObjects:[AnyObject]? = try context.executeFetchRequest(fetchRequest)
            
            //遍历查询的结果
            for info:Dog in fetchedObjects as! [Dog]{
                print("\nname=\(info.name)")
                print("sex=\(info.sex)")
                print("age=\(info.age)")
                
                info.sex = "女";
            }
            try context.save()
        }
        catch {
            fatalError("不能保存：\(error)")
        }
    }
    
    
    func clickBtnSelect() {
        //获取管理的数据上下文 对象
        let app = UIApplication.sharedApplication().delegate as! AppDelegate
        let context = app.managedObjectContext
        
        //声明数据的请求
        let fetchRequest:NSFetchRequest = NSFetchRequest()
        fetchRequest.fetchLimit = 10 //限定查询结果的数量
        fetchRequest.fetchOffset = 0 //查询的偏移量
        
        //声明一个实体结构
        let entity:NSEntityDescription? = NSEntityDescription.entityForName("Dog", inManagedObjectContext: context)
        //设置数据请求的实体结构
        fetchRequest.entity = entity
        
        //设置查询条件
//        let predicate = NSPredicate(format: "sex= '男' " )
//        fetchRequest.predicate = predicate
        
        //查询操作
        do {
            let fetchedObjects:[AnyObject]? = try context.executeFetchRequest(fetchRequest)
            
            if ( fetchedObjects?.count != 0 ) {
                //遍历查询的结果
                for info:Dog in fetchedObjects as! [Dog]{
                    print("\nname=\(info.name)")
                    print("sex=\(info.sex)")
                    print("age=\(info.age)")
                }
            }else{
                print( "没有查到数据" )
            }
            
            
        }
        catch {
            fatalError("不能保存：\(error)")
        }
    }
    
    
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


    
}

